api/relay.go

```diff
@@ -0,0 +1,306 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+	"sync"
+
+	"api.audius.co/api/dbv1"
+	"api.audius.co/config"
+	"api.audius.co/trashid"
+	"connectrpc.com/connect"
+	v1 "github.com/OpenAudio/go-openaudio/pkg/api/core/v1"
+	cconfig "github.com/OpenAudio/go-openaudio/pkg/core/config"
+	"github.com/OpenAudio/go-openaudio/pkg/core/server"
+	eth_gen "github.com/OpenAudio/go-openaudio/pkg/eth/contracts/gen"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	gcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/gofiber/fiber/v2"
+	"go.uber.org/zap"
+)
+
+const (
+	// ABI Function Arguments
+	manageEntityFunctionName      = "manageEntity"
+	manageEntityMetadataArgName   = "_metadata"
+	manageEntityActionArgName     = "_action"
+	manageEntityUserIdArgName     = "_userId"
+	manageEntityEntityTypeArgName = "_entityType"
+	manageEntityEntityIdArgName   = "_entityId"
+	manageEntityNonceArgName      = "_nonce"
+	manageEntitySubjectSigArgName = "_subjectSig"
+)
+
+var (
+	// entityManagerABI global singleton
+	entityManagerABI  *abi.ABI
+	entityManagerOnce sync.Once
+	entityManagerErr  error
+
+	// anonymouslyAllowedActions map that matches operation to struct{}
+	anonymouslyAllowedActions = map[string]struct{}{
+		"CreateUser":       {},
+		"DownloadTrack":    {},
+		"ViewNotification": {},
+	}
+)
+
+type RelayRequest struct {
+	EncodedABI string `json:"encodedABI"`
+}
+
+func DecodeManageEntityABI(encodedABI string) (*v1.ManageEntityLegacy, error) {
+	entityManagerOnce.Do(func() {
+		var err error
+		entityManagerABI, err = eth_gen.EntityManagerMetaData.GetAbi()
+		if err != nil {
+			entityManagerErr = err
+		}
+	})
+	if entityManagerErr != nil {
+		return nil, fmt.Errorf("failed to get entity manager ABI: %w", entityManagerErr)
+	}
+	data := gcommon.FromHex(encodedABI)
+	method, err := entityManagerABI.MethodById(data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get method by ID: %w", err)
+	}
+
+	if method.Name != manageEntityFunctionName {
+		return nil, fmt.Errorf("expected manageEntity function, got %s", method.Name)
+	}
+
+	params := make(map[string]interface{})
+	err = method.Inputs.UnpackIntoMap(params, data[4:])
```

Comment:
kinda surprised there isn't a struct unpacking version of this that could leverage like struct tag validation or something